homeassistant/components/zha/config_flow.py

```diff
@@ -53,48 +183,261 @@ async def async_step_user(self, user_input=None):
+
+            if strategy == FORMATION_FORM_NEW_NETWORK:
+                async with self._connect_zigpy_app() as app:
+                    await app.form_network()
+            elif strategy == FORMATION_REUSE_SETTINGS:
+                pass
+            elif strategy == FORMATION_RESTORE_MANUAL_BACKUP:
+                return await self.async_step_upload_manual_backup()
+            elif strategy == FORMATION_RESTORE_AUTOMATIC_BACKUP:
+                return await self.async_step_restore_automatic_backup()
+
+            return await self._async_create_radio_entity()
+
+        async with self._connect_zigpy_app() as app:
+            # Check if the stick has any settings and load them
+            try:
+                await app.load_network_info()
+            except NetworkNotFormed:
+                pass
+            else:
+                # Load the current info while we're connected, to save time
+                self._backups = app.backups.backups.copy()
+                self._current_settings = zigpy.backups.NetworkBackup(
+                    network_info=app.state.network_info,
+                    node_info=app.state.node_info,
+                )
+
+        strategies = []
+
+        # Check if we have any automatic backups *and* if the backups differ from
+        # the current radio settings (since restoring would be redundant)
+        if self._backups and any(
+            not backup.is_compatible_with(self._current_settings)
+            for backup in self._backups
+        ):
+            strategies.append(FORMATION_RESTORE_AUTOMATIC_BACKUP)
+
+        if self._current_settings is not None:
+            strategies.append(FORMATION_REUSE_SETTINGS)
+
+        strategies.append(FORMATION_RESTORE_MANUAL_BACKUP)
+        strategies.append(FORMATION_FORM_NEW_NETWORK)
+
+        # Pick a default. Forming a new network will always be an option but we would
+        # prefer to suggest something non-destructive if it's available.
+        suggested_strategy = [
+            strategy
+            for strategy in strategies
+            if strategy
+            in (
+                FORMATION_RESTORE_AUTOMATIC_BACKUP,
+                FORMATION_REUSE_SETTINGS,
+                FORMATION_FORM_NEW_NETWORK,
+            )
+        ][0]
+
+        schema = {
+            vol.Required(FORMATION_STRATEGY, default=suggested_strategy): vol.In(
+                strategies
+            ),
+        }
+
+        return self.async_show_form(
+            step_id="choose_formation_strategy",
+            data_schema=vol.Schema(schema),
+        )
+
+    def _parse_uploaded_backup(
+        self, uploaded_file_id: str
+    ) -> zigpy.backups.NetworkBackup:
+        """Read and parse an uploaded backup JSON file."""
+        with process_uploaded_file(self.hass, uploaded_file_id) as file_path:
+            contents = file_path.read_text()
+
+        return zigpy.backups.NetworkBackup.from_dict(json.loads(contents))
```

Comment:
What if it's not a valid backup file?

Reply:
Ah I see now that `async_step_upload_manual_backup` guards for `ValueError`. Does `NetworkBackup.from_dict` also raise that?

Reply:
It does. It will fail if the wrong JSON file is uploaded.